File: tests/python/test_arma_caster.py
import gc

import gelexy._test as t
import numpy as np
import pytest
from numpy.testing import assert_array_equal
from scipy import sparse


def test01_vec():
    a = np.array([1, 2, 3], dtype=np.int64)
    b = np.array([0, 1, 2], dtype=np.int64)
    c = np.array([1, 3, 5], dtype=np.int64)
    x = np.arange(10000, dtype=np.int64)
    af = np.float32(a)
    bf = np.float32(b)

    # Check call with dynamically sized arrays
    assert_array_equal(t.add_ivec(a, b), c)

    # Implicit conversion
    assert_array_equal(t.add_ivec(af, b), c)

    with pytest.raises(TypeError, match="incompatible function arguments"):
        t.add_ivec(a, bf)

    # Try with a big array. This will move the result to avoid a copy
    assert_array_equal(t.add_ivec(x, x), 2 * x)


def test02_mat():
    a = np.array([[1, 2], [3, 4]], dtype=np.int64, order="F")
    b = np.array([[0, 1], [2, 3]], dtype=np.int64, order="F")
    c = np.array([[1, 3], [5, 7]], dtype=np.int64, order="F")
    x = np.arange(10000).reshape(100, 100).astype(np.int64, order="F")

    af = np.float32(a)
    bf = np.float32(b)
    ac = np.array([[1, 2], [3, 4]], dtype=np.int64, order="C")
    bc = np.array([[0, 1], [2, 3]], dtype=np.int64, order="C")

    assert_array_equal(t.add_imat(a, b), c)

    # Implicit conversion type
    assert_array_equal(t.add_imat(af, b), c)
    # implicit conversion order
    assert_array_equal(t.add_imat(ac, b), c)

    with pytest.raises(TypeError, match="incompatible function arguments"):
        t.add_imat(a, bf)
        t.add_imat(a, bc)

    # Try with a big array. This will move the result to avoid a copy
    assert_array_equal(t.add_imat(x, x), 2 * x)


def test03_inplace_update():
    a = np.array([1, 2, 3], dtype=np.int64)
    b = np.array([[0, 1], [2, 3]], dtype=np.int64, order="F")

    c = np.array([99, 2, 3], dtype=np.int64)
    d = np.array([[99, 1], [2, 3]], dtype=np.int64, order="F")

    t.update_ivec(a)
    assert_array_equal(a, c)

    t.update_imat(b)
    assert_array_equal(b, d)


def test04_prop():
    for j in range(3):
        c = t.ClassWithArmaMember()
        ref = np.ones((2, 2))
        if j == 0:
            c.member = ref

        for i in range(2):
            member = c.member
            if j == 2 and i == 0:
                member[0, 0] = 10
                ref[0, 0] = 10
            assert_array_equal(member, ref)
            del member
            gc.collect()
            gc.collect()

        member = c.member
        assert_array_equal(c.member_ro_ref, ref)
        assert_array_equal(c.member_ro_copy, ref)
        del c
        gc.collect()
        gc.collect()
        assert_array_equal(member, ref)


def test05_init_assert():
    a = np.array([[1, 2, 3], [1, 2, 3]], dtype=np.float64, order="F")
    b = np.array([[0, 2, 3], [1, 2, 3]], dtype=np.float64, order="F")

    A = t.ClassInitFromPython(a)
    A.get_mat()[0, 0] = 0

    # assert a to make sure use same address
    assert_array_equal(a, b)

    del a
    gc.collect()
    gc.collect()
    assert_array_equal(A.get_mat(), b)


def test06_optional_class():
    a = np.array([[1, 2, 3], [1, 2, 3]], dtype=np.float64, order="F")
    b = np.array([[0, 2, 3], [1, 2, 3]], dtype=np.float64, order="F")

    A = t.ClassInitOptional(a)
    A.get_mat()[0, 0] = 0
    # assert a to make sure use same address
    assert_array_equal(a, b)

    A = t.ClassInitOptional()
    assert_array_equal(A.get_mat(), b)


def test07_sparse():
    a = np.array([[1, 3, 4], [0, 0, 0], [1, 0, 0]], dtype=np.float64, order="F")
    assert_array_equal(a, t.sparse().todense())

    b = a + 1
    a = sparse.csc_matrix(a)
    assert_array_equal(b, t.sparse_add(a))

File: tests/python/test_arma_caster.cpp
#include "gelex/python/dense.h"
#include "gelex/python/sparse.h"

#include <nanobind/nanobind.h>
#include <nanobind/stl/optional.h>
#include <armadillo>
#include <optional>

namespace nb = nanobind;
using namespace nb::literals;
using namespace arma;

namespace bind
{
namespace nb = nanobind;
using namespace nb::literals;

struct ClassWithArmaMember
{
    dmat member = dmat(2, 2, fill::ones);
    const dmat& get_member_ref() { return member; }
    const dmat get_member_copy() { return member; }
};

class ClassInitFromPython
{
   public:
    ClassInitFromPython(dmat mat) : mat_{std::move(mat)} {}
    dmat& get_mat() { return mat_; }

   private:
    dmat mat_;
};

class ClassInitOptional
{
   public:
    ClassInitOptional(std::optional<dmat> mat) : mat_{std::move(mat)} {}
    dmat& get_mat()
    {
        if (mat_)
        {
            return *mat_;
        }
        return mat_default_;
    }

   private:
    std::optional<dmat> mat_;
    dmat mat_default_{{0, 2, 3}, {1, 2, 3}};
};

void register_dense_test(nb::module_& m)
{
    m.def(
        "add_ivec",
        [](const ivec& a, const ivec& b) -> ivec { return a + b; },
        "a"_a,
        "b"_a.noconvert());
    m.def(
        "add_imat",
        [](const imat& a, const imat& b) -> imat { return a + b; },
        "a"_a,
        "b"_a.noconvert());
    m.def("update_ivec", [](ivec& a) { a(0) = 99.0; }, "a"_a.noconvert());
    m.def("update_imat", [](imat& a) { a(0, 0) = 99; }, "a"_a.noconvert());

    nb::class_<ClassWithArmaMember>(m, "ClassWithArmaMember")
        .def(nb::init<>())
        .def_prop_ro("member_ro_ref", &ClassWithArmaMember::get_member_ref)
        .def_prop_ro("member_ro_copy", &ClassWithArmaMember::get_member_copy)
        .def_rw("member", &ClassWithArmaMember::member);

    nb::class_<ClassInitFromPython>(m, "ClassInitFromPython")
        .def(nb::init<dmat>(), nb::keep_alive<1, 2>(), "mat"_a.noconvert())
        .def(
            "get_mat",
            &ClassInitFromPython::get_mat,
            nb::rv_policy::reference_internal);

    nb::class_<ClassInitOptional>(m, "ClassInitOptional")
        .def(
            nb::init<std::optional<dmat>>(),
            nb::keep_alive<1, 2>(),
            "mat"_a.noconvert() = nb::none())
        .def(
            "get_mat",
            &ClassInitOptional::get_mat,
            nb::rv_policy::reference_internal);
    m.def(
        "sparse",
        []() { return sp_dmat(dmat{{1, 3, 4}, {0, 0, 0}, {1, 0, 0}}); },
        nb::rv_policy::move);
    m.def(
        "sparse_add",
        [](sp_dmat a) -> dmat { return a + 1; },
        nb::rv_policy::move);
}
}  // namespace bind
NB_MODULE(_test, m)
{
    bind::register_dense_test(m);
}
